### PR TITLE
Use Remote State to generate traces for Debug API

### PR DESCRIPTION
**Pull request by m-Peter**
Closes: https://github.com/onflow/flow-evm-gateway/issues/530 https://github.com/onflow/flow-evm-gateway/issues/583

## Description

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-nft/blob/master/CONTRIBUTING.md#styleguides).
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: api/debug.go [hunk 1]
```diff
@@ -43,11 +70,93 @@ func (d *DebugAPI) TraceTransaction(
...
+	}
+
+	blockExecutor, err := evm.NewBlockExecutor(
+		previousBlock,
+		ledger,
+		d.config.FlowNetworkID,
+		d.blocks,
+		d.receipts,
+		d.logger,
+	)
```
Comment (sideninja) [inline api/debug.go:125]: we should extract all this logic into a tracer component. I don't like for API layer to have too much logic, it starts to get duplicated and it's harder to test in isolation. I suggest extracting the above into a tracer component in the services that just takes in a config of the tracer.

Comment (sideninja) [reply api/debug.go:125]: at that point we should also write unit tests with just the tracer component

File: models/block.go [hunk 2]
```diff
@@ -73,11 +72,13 @@ func (b *Block) Hash() (gethCommon.Hash, error) {
...
-			return block, nil
-		}
+		return nil, err
+	}
 
-		return nil, fmt.Errorf("failed to cadence decode block [%s]: %w", event.String(), err)
+	var fixedHash *string
+	if payload.PrevRandao == gethCommon.HexToHash("0x0") {
+		hash := payload.Hash.String()
+		fixedHash = &hash
```
Comment (ramtinms) [inline models/block.go:81]: not sure why we populate it using this heuristic?

Comment (m-Peter) [reply models/block.go:81]: This change is not really related to the objective of the PR, but it is required due to upgrading the `flow-go` versioned.
This heuristic allows us to decode block events prior to the addition of the `PrevRandao` field. The decoding of legacy block events was moved in flow-go (https://github.com/onflow/flow-go/blob/master/fvm/evm/events/events.go#L188-L190), so we no longer have a way to distinguish them, hence the usage of this heuristic. I'll add a comment to describe this though, as it is for sure not clear why it's needed :sweat_smile:

**Comment by m-Peter** [comment]
Closing in favor of: https://github.com/onflow/flow-evm-gateway/pull/635

**Pull request closed without merging**